# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1454530215 18000
#      Wed Feb 03 15:10:15 2016 -0500
# Node ID 82a44483dbffded7c4d8ebfad7fc908d9e8d30f7
# Parent  a65b906e25bc5c9a7871cc659d012b64972cd7dd
move octave_config_features function out of header file

* oct-conf-features.h: New file.
* module.mk: Generate octave-conf-features.cc instead of
oct-conf-features.h.
* toplev.cc (Foctave_config_info): Call octave::config::features.
* config-features.sh: Adapt to use of namespace.  Make internal data
static and initialized only once.

diff --git a/libinterp/config-features.sh b/libinterp/config-features.sh
--- a/libinterp/config-features.sh
+++ b/libinterp/config-features.sh
@@ -1,36 +1,57 @@
 #! /bin/sh
 
+## Attempt to get traditional sort behavior based on byte values.
+LC_ALL="C"
+export LC_ALL
+
 set -e
 AWK=${AWK:-awk}
 
 conffile=$1
 
 cat << EOF
 // DO NOT EDIT!  Generated automatically from $conffile by Make."
 
-#include "oct-map.h"
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "oct-conf-features.h"
 #include "ov.h"
 
-octave_scalar_map
-octave_config_features (void)
+namespace octave
 {
-  octave_scalar_map m;
+  namespace config
+  {
+    octave_scalar_map
+    features (void)
+    {
+      static bool initialized = false;
 
+      static octave_scalar_map m;
+
+      if (! initialized)
+        {
 EOF
 
 $AWK \
   '/#define (HAVE|ENABLE)_/ {
-     sub (/(HAVE|ENABLE)_/, "", $2);
-     printf ("  m.assign (\"%s\", octave_value (true));\n", $2);
+     sub (/HAVE_/, "", $2);
+     printf ("          m.assign (\"%s\", octave_value (true));\n", $2);
    }
    /\/\* #undef (HAVE|ENABLE)_/ {
-     sub (/(HAVE|ENABLE)_/, "", $3);
-     printf ("  m.assign (\"%s\", octave_value (false));\n", $3);
+     sub (/HAVE_/, "", $3);
+     printf ("          m.assign (\"%s\", octave_value (false));\n", $3);
    } {
-   }' $conffile
+   }' $conffile | sort
 
 cat << EOF
 
-  return m;
-}
+          initialized = true;
+        }
+
+      return m;
+    }
+  };
+};
 EOF
diff --git a/libinterp/corefcn/toplev.cc b/libinterp/corefcn/toplev.cc
--- a/libinterp/corefcn/toplev.cc
+++ b/libinterp/corefcn/toplev.cc
@@ -1451,17 +1451,17 @@ specified option.\n\
                 octave_value (oct_mach_info::float_format_as_string (ff)));
 
       m.assign ("words_big_endian",
                 octave_value (oct_mach_info::words_big_endian ()));
 
       m.assign ("words_little_endian",
                 octave_value (oct_mach_info::words_little_endian ()));
 
-      m.assign ("features", octave_value (octave_config_features ()));
+      m.assign ("features", octave_value (octave::config::features ()));
 
       int i = 0;
 
       while (true)
         {
           const conf_info_struct& elt = conf_info[i++];
 
           const char *key = elt.key;
diff --git a/libinterp/module.mk b/libinterp/module.mk
--- a/libinterp/module.mk
+++ b/libinterp/module.mk
@@ -31,17 +31,16 @@ octlib_LTLIBRARIES += libinterp/liboctin
 ## This is the subset of $(BUILT_SOURCES) that may be included by source
 ## files that are preprocessed to make $(DEF_FILES).  This ensures that
 ## files in $(BUILT_SOURCES) are built in the right dependency order.
 GENERATED_MAKE_BUILTINS_INCS = \
   libinterp/corefcn/defaults.h \
   libinterp/corefcn/graphics-props.cc \
   libinterp/corefcn/graphics.h \
   libinterp/corefcn/mxarray.h \
-  libinterp/oct-conf-features.h \
   libinterp/parse-tree/oct-gperf.h \
   libinterp/parse-tree/oct-parse.h \
   libinterp/version.h
 
 BUILT_SOURCES += \
   $(GENERATED_MAKE_BUILTINS_INCS) \
   libinterp/builtin-defun-decls.h \
   libinterp/builtins.cc \
@@ -76,17 +75,17 @@ LIBINTERP_BUILT_DISTFILES = \
 LIBINTERP_BUILT_NODISTFILES = \
   libinterp/corefcn/mxarray.h \
   libinterp/corefcn/oct-errno.cc \
   libinterp/corefcn/defaults.h \
   libinterp/corefcn/graphics.h \
   libinterp/builtin-defun-decls.h \
   libinterp/operators/ops.cc \
   libinterp/oct-conf.cc \
-  libinterp/oct-conf-features.h \
+  libinterp/oct-conf-features.cc \
   libinterp/version.h \
   $(OPT_HANDLERS) \
   $(ALL_DEF_FILES) \
   libinterp/builtins.cc
 
 libinterp_EXTRA_DIST += \
   libinterp/DOCSTRINGS \
   libinterp/config-features.sh \
@@ -102,16 +101,17 @@ libinterp_EXTRA_DIST += \
   $(LIBINTERP_BUILT_DISTFILES)
 
 octinclude_HEADERS += \
   libinterp/corefcn/graphics-props.cc \
   libinterp/parse-tree/oct-gperf.h \
   libinterp/builtins.h \
   libinterp/builtin-defun-decls.h \
   libinterp/oct-conf.h \
+  libinterp/oct-conf-features.h \
   libinterp/octave.h \
   libinterp/options-usage.h \
   $(OCTAVE_VALUE_INC) \
   $(PARSE_TREE_INC) \
   $(PARSER_INC) \
   $(LIBINTERP_OPERATORS_INC) \
   $(COREFCN_INC)
 
@@ -156,17 +156,17 @@ nodist_libinterp_liboctinterp_la_SOURCES
   libinterp/corefcn/mxarray.h \
   libinterp/corefcn/oct-errno.cc \
   libinterp/corefcn/defaults.h \
   libinterp/corefcn/graphics.h \
   libinterp/operators/ops.cc \
   libinterp/builtin-defun-decls.h \
   libinterp/builtins.cc \
   libinterp/oct-conf.cc \
-  libinterp/oct-conf-features.h \
+  libinterp/oct-conf-features.cc \
   libinterp/version.h
 
 libinterp_liboctinterp_la_LIBADD = \
   libinterp/octave-value/liboctave-value.la \
   libinterp/parse-tree/libparse-tree.la \
   libinterp/corefcn/libcorefcn.la \
   $(top_builddir)/liboctave/liboctave.la \
   $(LIBOCTINTERP_LINK_DEPS)
@@ -247,17 +247,17 @@ nobase_libinterptests_DATA = $(LIBINTERP
 
 ## oct-conf.cc must depend on Makefile.
 ## Calling configure may change default/config values.
 ## However, calling configure will also regenerate the Makefiles from
 ## Makefile.am and trigger the rules below.
 libinterp/oct-conf.cc: libinterp/oct-conf.in.cc Makefile
 	$(AM_V_GEN)$(do_subst_config_vals)
 
-libinterp/oct-conf-features.h: $(top_builddir)/config.h libinterp/config-features.sh
+libinterp/oct-conf-features.cc: $(top_builddir)/config.h libinterp/config-features.sh
 	$(AM_V_GEN)rm -f $@-t && \
 	$(srcdir)/libinterp/config-features.sh $< > $@-t && \
 	$(simple_move_if_change_rule)
 
 libinterp/version.h: libinterp/version.in.h Makefile
 	$(AM_V_GEN)rm -f $@-t && \
 	$(SED) < $< \
 	  -e "s|%NO_EDIT_WARNING%|DO NOT EDIT!  Generated automatically from $(<F) by Make.|" \
diff --git a/libinterp/oct-conf-features.h b/libinterp/oct-conf-features.h
new file mode 100644
--- /dev/null
+++ b/libinterp/oct-conf-features.h
@@ -0,0 +1,36 @@
+/*
+
+Copyright (C) 2016 John W. Eaton
+
+This file is part of Octave.
+
+Octave is free software; you can redistribute it and/or modify it
+under the terms of the GNU General Public License as published by the
+Free Software Foundation; either version 3 of the License, or (at your
+option) any later version.
+
+Octave is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received a copy of the GNU General Public License
+along with Octave; see the file COPYING.  If not, see
+<http://www.gnu.org/licenses/>.
+
+*/
+
+#if ! defined (octave_conf_features_h)
+#define octave_conf_features_h 1
+
+#include "oct-map.h"
+
+namespace octave
+{
+  namespace config
+  {
+    extern OCTAVE_API octave_scalar_map features (void);
+  };
+};
+
+#endif
